### PR TITLE
Add small test for nova evacuate.

### PR DESCRIPTION
**Pull request by gosipyan**

### CODE REVIEW

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 1]
```diff
@@ -179,6 +179,60 @@
...
+          echo $hypervisor_host
+          export hypervisor_host=$hypervisor_host
+          # Crate instance on specific hypervisor image always exist after tempest run (cirros-0.3.4-x86_64-tempest-machine)
+          nova boot --image cirros-0.3.4-x86_64-tempest-machine  --flavor  tempest-stuff  --availability-zone nova:$hypervisor_host NovaEvacuate
+
+          #Check instance is ACTIVE and Running
+          status=$(nova show  NovaEvacuate | grep status | cut -d '|' -f 3 | grep ACTIVE)
+          echo $status
+          if [ $status = "ACTIVE" ] ; then
+           echo "Instance launched successfully"
```
Comment (aplanas) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: two spaces indentation?

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 2]
```diff
@@ -179,6 +179,60 @@
...
+          else
+           echo  "Instance luanch failed"
```
Comment (aplanas) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: same hre

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 3]
```diff
@@ -179,6 +179,60 @@
...
+          fi
+          EOF
+          )
+          hypervisor_host=$(echo $OUTPUT | cut -d ' ' -f 1)
+
+          #Get full hypervisor hostname
+          echo "$hypervisor_host"  > temp
```
Comment (aplanas) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: I understand why you are using it, and I really do not know a better approach here. Can you use `hypervisor_host` as a file name to be more clear? also there is an extra space.

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 4]
```diff
@@ -179,6 +179,60 @@
...
+          hypervisor_fqdn=$(crowbarctl node list | grep $hypervisor_host | cut -d '|' -f 2)
+          scp temp controller1:/root/
```
Comment (aplanas) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: I would do the scp after the creation of the file

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 5]
```diff
@@ -179,6 +179,60 @@
...
+          #power off hypervisor node
```
Comment (aplanas) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: is a reboot more that a power off, I guess

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 6]
```diff
@@ -179,6 +179,60 @@
...
+          crowbarctl node reboot $hypervisor_fqdn
+          #wait
+          sleep 180
```
Comment (aplanas) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: can you try to wait until the ssh answer? I think that there is code somewhere (mkcloud?) that wait until ssh is ready, or there is a bit timeout

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 7]
```diff
@@ -179,6 +179,60 @@
...
+
+          #Check that instance evacuated
+
+          ssh -T controller1  << 'EOF'
+
+          hypervisor_host=$(cat /root/temp)
```
Comment (aplanas) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: if you decide to change the name, you need to change it here too

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 8]
```diff
@@ -179,6 +179,60 @@
...
+          source /root/.openrc
+          #Check that instance evacuated to working hypervisor host
+          hypervisor_evacuate_host=$(nova show  NovaEvacuate | grep hypervisor_hostname | cut -d '|' -f 3 | cut -d '.' -f 1 |cut -d ' ' -f 2)
```
Comment (aplanas) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: extra space before NovaEvacuate, there are also too many cut -f, can you write a text example of the output in the comment?

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 9]
```diff
@@ -179,6 +179,60 @@
...
+          if [ "$hypervisor_evacuate_host" != "$hypervisor_host" ] ; then
+          #Check that instance ACTIVE and Running
+          status=$(nova show  NovaEvacuate | grep status | cut -d '|' -f 3 | grep ACTIVE)
+          if [ $status = "ACTIVE" ] ; then
+           echo "Instance evacuated"
```
Comment (aplanas) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: two indentation spaces (also later)

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 10]
```diff
@@ -155,18 +155,18 @@
...
-          export UPDATEBEFOREINSTALL=$UPDATEBEFOREINSTALL ;
-          export TESTHEAD=$TESTHEAD ;
-          export cloudsource=$cloudsource ;
-          export nodenumber=$nodenumber ;
-          export hacloud=$hacloud ;
-          export clusterconfig=$clusterconfig ;
-          export shared_storage_ip=$shared_storage_ip ;
-          export want_node_aliases=controller=3,compute=3;
-          export scenario=\"/root/scenario.yml\" ;
+          export cloud=$cloud
```
Comment (bmwiedemann) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: such cleanup should be a separate commit. you can use `git reset --patch HEAD^` and `git commit --amend` to split things out.

Comment (aplanas) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: IMHO for this kind of document is OK to have it here, but I do not mind either

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 11]
```diff
@@ -179,6 +179,66 @@
...
+          status=$(nova show  NovaEvacuate | grep status | cut -d '|' -f 3 | grep ACTIVE)
+          echo $floating_ip
+          if [ $status = "ACTIVE" ] && [ $? = 0 ] ; then
+          echo "Instance launched successfully"
+          else
+          echo  "Instance luanch failed"
+          fi
+          EOF
+          )
+          hypervisor_host=$(echo $OUTPUT | cut -d ' ' -f 1)
```
Comment (aspiers) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: Isn't this going to catch output from multiple lines of `$OUTPUT`?  What about if the `nova boot` fails, or any of the other commands here fail?  We need the whole of `$OUTPUT` visible in the build logs so that we can easily debug what went wrong.  So at least there should be an `echo -e "--- Starting $vm_name on controller1 ---\n$OUTPUT\n--- End of output ---"` or similar.

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 12]
```diff
@@ -179,6 +179,66 @@
...
+          floating_ip=$(echo $OUTPUT | cut -d ' ' -f 2)
+          #Get full hypervisor hostname
+          echo "$hypervisor_host"  > hypervisor_host
+          scp hypervisor_host controller1:/root/
+          hypervisor_fqdn=$(crowbarctl node list | grep $hypervisor_host | cut -d '|' -f 2)
+          #Reboot  hypervisor node
+          ssh $hypervisor_fqdn "killall -9 acemaker_remoted"
```
Comment (aspiers) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: I don't think it's necessary to drop the `p` from `pacemaker_remoted` here.

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 13]
```diff
@@ -179,6 +179,66 @@
...
+          fi
+          EOF
+          )
+          hypervisor_host=$(echo $OUTPUT | cut -d ' ' -f 1)
+          floating_ip=$(echo $OUTPUT | cut -d ' ' -f 2)
+          #Get full hypervisor hostname
+          echo "$hypervisor_host"  > hypervisor_host
+          scp hypervisor_host controller1:/root/
+          hypervisor_fqdn=$(crowbarctl node list | grep $hypervisor_host | cut -d '|' -f 2)
+          #Reboot  hypervisor node
```
Comment (aspiers) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: This comment should actually say `# Simulate failure of hypervisor node`, since it's different from a (clean) reboot - the node will get fenced.

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 14]
```diff
@@ -179,6 +179,66 @@
...
+          #Update security group for icmp 
+          nova secgroup-add-rule default icmp -1 -1 0.0.0.0/0
+          #Assign public ip to instance
+          nova add-floating-ip NovaEvacuate  $floating_ip
+          #Check instance is ACTIVE and Running
+          ping -c 5 $floating_ip >/dev/null 2>&1
+          status=$(nova show  NovaEvacuate | grep status | cut -d '|' -f 3 | grep ACTIVE)
+          echo $floating_ip
+          if [ $status = "ACTIVE" ] && [ $? = 0 ] ; then
+          echo "Instance launched successfully"
```
Comment (aspiers) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: Some extra indentation would be nice here.

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 15]
```diff
@@ -179,6 +179,66 @@
...
+          scp hypervisor_host controller1:/root/
+          hypervisor_fqdn=$(crowbarctl node list | grep $hypervisor_host | cut -d '|' -f 2)
+          #Reboot  hypervisor node
+          ssh $hypervisor_fqdn "killall -9 acemaker_remoted"
+          ping -c 5 $hypervisor_fqdn >/dev/null 2>&1
+          echo $?
+          #wait
+          n=300
+          while test $n -gt 0 && ! ping -q -c 1 -w $floating_ip >/dev/null  2>&1 ; do
+          n=$(expr $n - 1)
```
Comment (aspiers) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: Some extra indentation would be nice here too.

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 16]
```diff
@@ -179,6 +179,66 @@
...
+
+          #Check that instance evacuated
+          ssh -T controller1  << 'EOF'
+
+          hypervisor_host=$(cat /root/hypervisor_host)
+          source /root/.openrc
+          #Check that instance evacuated to working hypervisor host
+          hypervisor_evacuate_host=$(nova show NovaEvacuate | grep hypervisor_hostname | cut -d '|' -f 3 | cut -d '.' -f 1 |cut -d ' ' -f 2)
+          if [ "$hypervisor_evacuate_host" != "$hypervisor_host" ] ; then
+          echo "Instance evacuated"
```
Comment (aspiers) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: Some extra indentation would be nice here too.

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 17]
```diff
@@ -179,6 +179,66 @@
           timeout --signal=ALRM 150m bash -x -c ". scripts/qa_crowbarsetup.sh ; onadmin_runlist $commands;"
           ' || ret=$?
 
+          #Test nova evacuate
+          OUTPUT=$(ssh -T controller1 <<'EOF' 
+          source /root/.openrc
```
Comment (aspiers) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: Would be nicer to indent the whole of the `EOF` heredoc more than its surrounding `EOF` delimiters.

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 18]
```diff
@@ -178,7 +183,22 @@
 
           timeout --signal=ALRM 150m bash -x -c ". scripts/qa_crowbarsetup.sh ; onadmin_runlist $commands;"
           ' || ret=$?
-
+          
+
+          if [ $RUNNOVAEVACUATE == "true" ] && [ $ret=0 ] ; then
```
Comment (bmwiedemann) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: `[ $ret=0 ]` is broken. must be [[ $ret = 0 ]] or [ "$ret" = 0 ]
```
( ret=1 ; [ $ret=0 ] ; echo $?)
0
```

File: scripts/qa_crowbarsetup.sh [hunk 19]
```diff
@@ -3343,6 +3343,89 @@ function addfloatingip
     nova floating-ip-create | tee floating-ip-create.out
     floatingip=$(perl -ne "if(/\d+\.\d+\.\d+\.\d+/){print \$&}" floating-ip-create.out)
     nova add-floating-ip "$instanceid" "$floatingip"
+    echo $floatingip
```
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: why add this echo?

File: scripts/qa_crowbarsetup.sh [hunk 20]
```diff
@@ -3343,6 +3343,89 @@ function addfloatingip
...
+    # Update security group for icmp
+    nova secgroup-add-rule default icmp -1 -1 0.0.0.0/0 &> /dev/null
+
+    # Check if instance is ACTIVE and Running
+    ping -c 10 $floatingip >/dev/null 2>&1
+    local ping_result=$?
+    echo "Instance ping res" $ping_result
+    local status=$(openstack server show -c status --format value $vm_name)
+    echo $status
+    if [ "$status" == "ACTIVE" ] && [ $ping_resul=0 ] ; then
```
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: same problem here. and typo in `resul`

File: scripts/qa_crowbarsetup.sh [hunk 21]
```diff
@@ -3343,6 +3343,89 @@ function addfloatingip
...
+    tail -f -n0 $MSG_LOG | grep 'Completed evacuation of'
+    ret=$?
+    n=$(expr $n - 1)
+    done
+    echo "Hypervisor killed"
+    # Wait to instance be up
+    local n=300
+    while test $n -gt 0 && ! ping -q -c 1 -w $floatingip >/dev/null 2>&1 ; do
+        n=$(expr $n - 1)
+        echo -n "."
```
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: loop could probably be replaced by a call to wait_for

File: scripts/qa_crowbarsetup.sh [hunk 22]
```diff
@@ -3343,6 +3343,89 @@ function addfloatingip
...
+
+    # Simulate failure of hypervisor node
+    echo $hypervisor_fqdn
+    ssh -t -T  $hypervisor_fqdn 'killall -9 pacemaker_remoted'
+    local MSG_LOG="/var/log/messages"
+    local ret=1
+    local n=300
+    while [ $n -gt 0 ] && [ "$ret" != "0" ]; do
+    #cat $MSG_LOG | grep -qe $SEARCH_PATTERN
+    tail -f -n0 $MSG_LOG | grep 'Completed evacuation of'
```
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: tail -f will not return, unless you used grep -q behind the pipe which might exit on first match and then close, which triggers a SIGPIPE in tail.

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml [hunk 23]
```diff
@@ -185,7 +185,7 @@
           ' || ret=$?
           
 
-          if [ $RUNNOVAEVACUATE == "true" ] && [ $ret=0 ] ; then
+          if [ $RUNNOVAEVACUATE == "true" ] && [ "$ret"=0 ] ; then
```
Comment (bmwiedemann) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: still needs spaces around = operator

File: scripts/qa_crowbarsetup.sh [hunk 24]
```diff
@@ -3378,7 +3378,7 @@ function oncontroller_nova_evacuate
     echo "Instance ping res" $ping_result
     local status=$(openstack server show -c status --format value $vm_name)
     echo $status
-    if [ "$status" == "ACTIVE" ] && [ $ping_resul=0 ] ; then
+    if [ "$status" == "ACTIVE" ] && [ $ping_result=0 ] ; then
```
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: same here

File: scripts/qa_crowbarsetup.sh (hunk 19 above)
Comment (gosipyan) [inline scripts/qa_crowbarsetup.sh]: @bmwiedemann I do it to get flotingip when I call function in my script

File: jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml (hunk 10 above)
Comment (gosipyan) [inline jenkins/ci.suse.de/cloud-mkphyscloud-qa-scenario-3.yaml]: done

**Comment by gosipyan** [comment]
Test successfully run and pass on Jenkins.
https://ci.suse.de/view/Cloud/view/QA/job/cloud-mkphyscloud-qa-scenario-3/48/console
See after: 21:16:47 Start nova evacuate testing

File: scripts/qa_crowbarsetup.sh [hunk 25]
```diff
@@ -3346,6 +3346,90 @@ function addfloatingip
...
+    local ret=1
+    local n=120
+    # Check for nova-evacute log on all cluster nodes
+    while [[ $n -gt 0 && $ret != 0 ]]; do
+        for((i=1; i<=$num_controllers; i++)) ; do
+            ssh controller$i 'grep -q "Completed evacuation of" /var/log/messages'
+            ret=$?
+            [[ $ret != 0 ]] && break
+        done
+    n=$(expr $n - 1)
```
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: wrong indention.
this loop looks a bit like a "wait_for" but it is missing a "sleep" so it is unclear how long you are waiting.

Comment (gosipyan) [inline scripts/qa_crowbarsetup.sh]: Fixed

File: scripts/qa_crowbarsetup.sh [hunk 26]
```diff
@@ -3354,6 +3354,91 @@ function addfloatingip
...
+        for((i=1; i<=$num_controllers; i++)) ; do
+            ssh controller$i 'grep -q "Completed evacuation of" /var/log/messages'
+            ret=$?
+            [[ $ret != 0 ]] && break
+        done
+        sleep 1
+        n=$(expr $n - 1)
+    done
+    echo "Wait to instance be up"
+    # Wait to instance be up
```
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: comment is redundant. grammar/typo errors above and below s/Instanse/instance/". And wait_for will already print such a message.

File: scripts/qa_crowbarsetup.sh [hunk 27]
```diff
@@ -3354,6 +3354,91 @@ function addfloatingip
...
+function kill_pacemaker_remote
+{
+    echo "Killing pacemaker..."
+
+    local hypervisor_fqdn=$1
+    local floatingip=$2
+    local num_controllers=$3
+
+    # Simulate failure of hypervisor node
+    ssh -T $hypervisor_fqdn "killall -9 pacemaker_remoted"
```
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: I still think, that this is not testing some real use-case. hypervisor failure would be "reboot -f"

File: scripts/qa_crowbarsetup.sh [hunk 28]
```diff
@@ -3354,6 +3354,91 @@ function addfloatingip
...
+
+    # JSON parser with python
+    local PJSON='import json,sys; field=json.load(sys.stdin)'
+
+    # Get hypervisor hostname
+    local hypervisors=$(openstack hypervisor list -c 'Hypervisor Hostname' --format json)
+    local hypervisor_host_fqdn=$(echo $hypervisors | python -c "$PJSON; print field[0]['Hypervisor Hostname']")
+    local hypervisor_host=$(echo $hypervisor_host_fqdn | cut -d '.' -f 1)
+
+    # Crate instance on specific hypervisor image always exist after tempest
```
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: s/Crate/Create/
and some grammar

File: scripts/qa_crowbarsetup.sh (hunk 26 above)
Comment (gosipyan) [inline scripts/qa_crowbarsetup.sh]: Fixed

File: scripts/qa_crowbarsetup.sh (hunk 28 above)
Comment (gosipyan) [inline scripts/qa_crowbarsetup.sh]: Fixed

File: scripts/qa_crowbarsetup.sh (hunk 27 above)
Comment (gosipyan) [inline scripts/qa_crowbarsetup.sh]: @bmwiedemann this was recommended by @aspiers .And reboot was wrong option as far as I get from him.

**Comment by aplanas** [comment]
@bmwiedemann you are right, a better approach in needed, but is only 80 lines in qa_crowbarsetup.sh

I think that Rally can replace most if not all the tests we have that needs to interact with the cloud

**Comment by gosipyan** [comment]
@bmwiedemann agree, will remove after integrating rally and creating similar test there.

File: scripts/qa_crowbarsetup.sh (hunk 27 above)
Comment (aspiers) [inline scripts/qa_crowbarsetup.sh]: `reboot -f` should also be a valid test.  `reboot` *without* `-f` would not.  But the current test is fine.
